### PR TITLE
added ValueEquals interface to test value equivalency

### PR DESCRIPTION
**Pull request by axtimwalde**
added ValueEquals interface to test value equivalency,
made Type extend this interface and implemented it across all core types,
bumped version to 3.0.0-SNAPSHOT

### CODE REVIEW

**Comment by axtimwalde** [comment]
Note that this change requires to raise the major version.  It is possible ti increase only the minor version by not having `Type extends ValueEquals` and have only each leaf implementation of `Type` implement this interface individually.  Implementations at the abstract class level would again require to raise the major version. I do not like this option and would prefer the major version increase.

**Comment by ctrueden** [comment]
What if you add a `default` method to the interface which simply returns false?

Before Java 8, our policy at the ImageJ2 layer was that if we added a new method to an interface, but added a default implementation to the corresponding abstract class(es), then we did not bump the major version.

There is a difference between __API__ compatibility—whether downstream _calling code_ would break—and __SPI__ compatibility—whether downstream _extending code_ would break. We opt to have SemVer mean __API__ compatibility, although we make a best effort for SPI compatibility as well as described above.

**Comment by axtimwalde** [comment]
@ctrueden, that would make imglib2 depend on Java8 for which there is currently no reason.  Regardless of this, I do not like the default implementation here for that it hides undesired behavior from unaware consumers which is generally a bad thing.

**Comment by axtimwalde** [comment]
ImgLib2's 'API' being merely interfaces I also strongly vote for SPI compatibility.

**Comment by ctrueden** [comment]
> that would make imglib2 depend on Java8 for which there is currently no reason.

Sorry, I was going on the assumption that ImgLib2 would switch to Java 8 imminently, as indicated by @tpietzsch in chat earlier today. Apologies if I misunderstood him.

> ImgLib2's 'API' being merely interfaces I also strongly vote for SPI compatibility.

Sure. I think we should document this on a per-component (or per-org) basis here somewhere.

**Comment by axtimwalde** [comment]
@ctrueden thanks for clarification.  I am not aware of an ImgLib2 -> Java8 discussion in a chat today.  Will wait for @tpietzsch's comments on this.

**Comment by tpietzsch** [comment]
@ctrueden @axtimwalde The imglib2 and imglib2-algorithm artefacts I released today are java8. I saw no reason to delay this any longer. I see no point of sticking with java6. We shouldn't deny ourselves the new java8 features and support a java version that is itself not supported anymore.

Regarding the original discussion: I didn't look at the PR yet. I'll comment tomorrow.

**Comment by axtimwalde** [comment]
@tpietzsch I agree in principle to go for Java8 whenever needed.  Is there Java8 code in imglib2 and imglib2-algorithms?  If not, we unnecessarily prevent them to be used in pre Java 8 environments.  Such an environment could be Mac OS or Ubuntu 15.04 LTS.

**Comment by axtimwalde** [comment]
@ctrueden interface heavy, i.e. interfaces make a big part of the API.  The second argument that the suggested default would generate undesired surprises is stronger.  Let's see what @tpietzsch has to say.

**Comment by tpietzsch** [comment]
@axtimwalde There is no java8 code, yet. However, there might be in the future. There will never really be a strong *need* to go Java8 because for every individual feature we want to have we can work around it with java6. I mean stuff like the need to have both `copy()` and `copyCursor()`. At least that holds for the core. That could in principle be java6 forever.

For other imglib2 components it will be more difficult. For components that use 3rd party libraries, it will mean that we cannot use the most recent versions of those libraries. This will eventually lead to clashes with other things in Fiji which require the more recent version, and then we or someone on the Fiji side will need to work around _that_.

This is about developer convenience vs user convenience. It is certainly possible to install Java 8 on Ubuntu 14.04 LTS and older, it's just more inconvenient for the user. At some point developer convenience outweighs user convenience. One can argue whether we are at that point already, but I think we should move on now.

File: src/main/java/net/imglib2/view/composite/AbstractNumericComposite.java [hunk 1]
```diff
@@ -193,4 +193,21 @@ public void mul( final double c )
...
 		return iterator;
 	}
+
+	@Override
+	public boolean valueEquals( C t )
+	{
+		if ( t.length == length )
+		{
+			final Iterator< T > a = iterator();
+			final Iterator< T > b = iterator();
```
Comment (tpietzsch) [inline src/main/java/net/imglib2/view/composite/AbstractNumericComposite.java]: One of these should be `t.iterator()`.

**Comment by tpietzsch** [comment]
Phew. This is a tough one.

I'm not sure whether ValueEquals should be a superclass of Type. I would tend to either keep it separate, or add the valueEquals method directly to the type interface.

I'm on the fence about the default method. On the one hand, it would be nice to have, because it wouldn't break downstream code. On the other hand, we might _actually want to break downstream code_ to make implementors of `Type` think about what to do. Actually I tend to doing the later thing. Then maybe a few releases down the line we can add the default method anyway. I would not have it return false but call `equals`.

About the version, I honestly don't care. Why not bump to 3.0.0? Who cares...

File: src/main/java/net/imglib2/view/composite/AbstractNumericComposite.java (hunk 1 above)
Comment (axtimwalde) [inline src/main/java/net/imglib2/view/composite/AbstractNumericComposite.java]: Yes, fixing it.

**Comment by axtimwalde** [comment]
Ok, version topic is settled, bump to 3.0.0 if we make it superclass of `Type`.

Regarding the superclass issue.  I was thinking about it for some time and could not find an example to have any `Type` not implement this interface but lots of examples where `ValueEquals` would not be a `Type`.  I think keeping it a separate interface would also be even less strict.  If my observation is true and all `Type` should implement `ValueEquals` superclassing `Type` is more convenient (saving some & in generics).  I will make a version that separates it from `Type` and see how that looks in comparison.

**Comment by tpietzsch** [comment]
@axtimwalde don't bother. Let's leave it like this if you think that `ValueEquals` is useful outside `Type`, or put the method directly into type is it's not.

**Comment by axtimwalde** [comment]
@tpietzsch, just started looking at it.  If I were to not superclass `Type`, I do not know where to start superclassing.  The next level is `BasePairType`, `BooleanType`, `NativeType`, `NumericType`, all good candidates but with the same consequence of introducing something new into interfaces.  The next level seems even more arbitrary, so I would go for implementing it only in leaves which I do not like for how messy and unstructured it is.  So I go for the 'don't bother', fix the bug and stick with how it is now.  `ValueType` as a separate interface makes sense e.g. for `Neighborhood` or other `IterableInterval`s.

**Comment by tpietzsch** [comment]
@axtimwalde ok, great. `Neighborhood` makes sense, yes. Good example

**Comment by ctrueden** [comment]
> we unnecessarily prevent them to be used in pre Java 8 environments. Such an environment could be Mac OS or Ubuntu 15.04 LTS.

Well, Ubuntu can run Java 8 since at least 12.04 LTS. We ship Java 8 with Fiji and AFAIK it runs fine on older Ubuntus. The main problem is OS X 10.7 Lion and earlier. And Windows 2000 maybe? 💻 

https://docs.google.com/spreadsheets/d/19rSofTznt8XT6hRUPKmDspayUYdSQujfTjVm1NqYQP4/edit?usp=sharing

**Pull request merged**